### PR TITLE
Call Cython via `python -m Cython` rather than system-wide binary

### LINKED ISSUE #1885: TypeError: must be str, not bytes
### Versions

* Python: 3.5.3 (running in virtualenv and in root)
  (seems to be running Python 3.7.1 on Android based on the logs)
* OS: Ubuntu 17:04
* Kivy: 1.9.1
* Cython: 0.28.6
* Andoid 7.0 (on Samsung Galaxy S6)

### Description

I have compiled a very basic helloWorld app as below. The app works fine on Ubuntu and also compiles fine into apk using buildozer. 

````
__version__ = '1.0'
from kivy.app import App
from kivy.uix.button import Button
 
class Hello(App):
    def build(self):
        btn = Button(text='Hello World')
        return  btn
 
Hello().run()
````

When I run on Android the app immediately closes before displaying the hello world screen. After working through logs produced with adb logcat I see the following...
````
06-25 19:48:52.272  2100  2118 I python  :  Traceback (most recent call last):
06-25 19:48:52.272  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/app/main.py", line 3, in <module>
06-25 19:48:52.273  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/platform/build/build/python-installs/helloWorld/kivy/app.py", line 319, in <module>
06-25 19:48:52.273  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/platform/build/build/python-installs/helloWorld/kivy/base.py", line 26, in <module>
06-25 19:48:52.274  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/platform/build/build/python-installs/helloWorld/kivy/clock.py", line 410, in <module>
06-25 19:48:52.274  2100  2118 I python  :    File "<frozen importlib._bootstrap>", line 983, in _find_and_load
06-25 19:48:52.274  2100  2118 I python  :    File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
06-25 19:48:52.275  2100  2118 I python  :    File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
06-25 19:48:52.275  2100  2118 I python  :    File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
06-25 19:48:52.276  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/platform/build/build/other_builds/python3-libffi-openssl-sqlite3/armeabi-v7a__ndk_target_21/python3/Lib/ctypes/util.py", line 73, in <module>
06-25 19:48:52.276  2100  2118 I python  :    File "/home/lee/KivyProjects/helloWorld/.buildozer/android/platform/build/build/python-installs/helloWorld/android/__init__.py", line 8, in <module>
06-25 19:48:52.277  2100  2118 I python  :    File "android/_android.pyx", line 162, in init android._android
06-25 19:48:52.277  2100  2118 I python  :  TypeError: must be str, not bytes
06-25 19:48:52.277  2100  2118 I python  : Python for android ended.
````

This appears to be the same issue raised as 1691 under the title "unicode error during startup (python3, numpy, opencv) - patch included #1691". If I am reading this patch correctly the current file _android.pyx should have the following changes

removed this line
    python_act = autoclass(JAVA_NAMESPACE + u'.PythonActivity')

added these lines
    java_namespace = JAVA_NAMESPACE.decode() if isinstance(JAVA_NAMESPACE, bytes) else JAVA_NAMESPACE
    python_act = autoclass(java_namespace  + u'.PythonActivity')

Yet when I look at the current code on Github for _android.pyx I do not see these changes applied. Has this fix somehow been removed from the current version? How can i apply this fix to my system?

### PR DESCRIPTION
**Pull request by JonasT**
This changes calls to Cython to `python -m Cython` to avoid picking one not matching the current python version (which can happen if just calling `Cython`). This should fix #1885

### CODE REVIEW

File: pythonforandroid/build.py [hunk 1]
```diff
@@ -374,7 +365,6 @@ def prepare_build_environment(self,
         if not self.ccache:
             info('ccache is missing, the build will not be optimized in the '
                  'future.')
-        self.cython = get_cython_path()
```
Comment (inclement) [inline pythonforandroid/build.py]: Could we still add a check here for Cython being available, or a BuildInterruptingException if not?

Comment (JonasT) [reply pythonforandroid/build.py]: Oh, good point! Sure, that would be possible, let me have a look at it

File: Dockerfile.py3 [hunk 2]
```diff
@@ -107,6 +107,7 @@ RUN dpkg --add-architecture i386 \
         libpangox-1.0-0:i386 libpangoxft-1.0-0:i386 libidn11:i386 \
         zip zlib1g-dev zlib1g:i386 \
     && apt -y autoremove
+RUN ${RETRY} pip3 install -U Cython
```
Comment (AndreMiras) [inline Dockerfile.py3]: Wait what? It's already there https://github.com/kivy/python-for-android/blob/v2019.07.08/Dockerfile.py3#L126

File: pythonforandroid/build.py [hunk 3]
```diff
@@ -374,7 +365,14 @@ def prepare_build_environment(self,
...
+        try:
+            subprocess.check_output([
+                "python3", "-m", "cython", "--help",
+            ])
+        except subprocess.CalledProcessError:
+            warning('Cython for python3 missing. If you are building for '
+                    ' a python 3 target (which is the default)'
+                    ' then THINGS WILL BREAK.')
```
Comment (AndreMiras) [inline pythonforandroid/build.py]: If things would definitely break. Then let's not be gentle and let's fail fast with `BuildInterruptingException`, what do you guys think?

File: pythonforandroid/recipe.py [hunk 4]
```diff
@@ -1024,8 +1024,12 @@ def cythonize_file(self, env, build_dir, filename):
             del cyenv['PYTHONPATH']
         if 'PYTHONNOUSERSITE' in cyenv:
             cyenv.pop('PYTHONNOUSERSITE')
-        cython_command = sh.Command(self.ctx.cython)
-        shprint(cython_command, filename, *self.cython_args, _env=cyenv)
+        cython_command = sh.Command(
+            "python" +
+            self.ctx.python_recipe.major_minor_version_string.split(".")[0]
```
Comment (AndreMiras) [inline pythonforandroid/recipe.py]: nitpick: I prefer `format()`

File: pythonforandroid/recipe.py [hunk 5]
```diff
@@ -1024,8 +1024,12 @@ def cythonize_file(self, env, build_dir, filename):
             del cyenv['PYTHONPATH']
         if 'PYTHONNOUSERSITE' in cyenv:
             cyenv.pop('PYTHONNOUSERSITE')
-        cython_command = sh.Command(self.ctx.cython)
-        shprint(cython_command, filename, *self.cython_args, _env=cyenv)
+        cython_command = sh.Command(
```
Comment (AndreMiras) [inline pythonforandroid/recipe.py]: technically this isn't a cython command, but a python command, correct?
The cython however comes later once you adds the `-m cython` to it few lines below

File: pythonforandroid/build.py (hunk 3 above)
Comment (JonasT) [reply pythonforandroid/build.py]: Only if it's a python 3 target. We still support python 2, and at this stage the recipes aren't known :woman_shrugging: so as the text says, it would break but only for a python 3 build. I did some poking around the code and I found no way to tell this in this function

Comment (AndreMiras) [reply pythonforandroid/build.py]: Oh Python2, right... Thanks I almost forgot about this folk :joy:

File: Dockerfile.py3 [hunk 6]
```diff
@@ -133,4 +135,5 @@ USER ${USER}
 # install python-for-android from current branch
 RUN virtualenv --python=python3 venv \
     && . venv/bin/activate \
+    && pip3 install -U Cython \
```
Comment (AndreMiras) [inline Dockerfile.py3]: Isn't the one installed system wide enough?
https://github.com/kivy/python-for-android/blob/v2019.07.08/Dockerfile.py3#L126

Comment (JonasT) [reply Dockerfile.py3]: Hm true I may have interpreted that one error situation wrong, let me try it might work without that :+1:

Comment (JonasT) [reply Dockerfile.py3]: Yup system-wide is enough, apparently :joy: well-spotted!

Comment (AndreMiras) [reply Dockerfile.py3]: I understand, thanks for investigating!
Then maybe we should ditch the system one?

Comment (AndreMiras) [reply Dockerfile.py3]: Ouch, I just saw the build is failing for Python2.
https://travis-ci.org/kivy/python-for-android/jobs/567481453
```
2943  RAN: /usr/bin/python2 -m Cython.Build.Cythonize ./jnius/jnius.pyx
2944
2945  STDOUT:
2946/usr/bin/python2: No module named Cython.Build
```
So that's properly weird and annoying. Basically it means if we want to support the "Python2 basic" apk build in the CI, we do need the system cython on python2. While only having the venv cython3. That would look kinda inconsistent in the Dockerfile, but with some comments that should make it.
So basically we want to have:
```
RUN pip2 install --upgrade Cython==0.28.6
```
And also in the venv:
```
&& pip3 install --upgrade Cython==0.28.6
```
By the way, let's take the opportunity to make them consistent.

- uppercase `Cython` because that's it's naming on pypi
- long argument `--upgrade` because we love explicit
- explicit "officially p4a tested" version `0.28.6` in all places (eventually via variable)

**Comment by JonasT** [comment]
I just realized I never tested if this *actually* fixes #1885, even though I believe this is the "more correct" way to call it in any case. Does anyone have a test environment for that around? Otherwise I suppose I should set myself one up real quick and actually give that a try

File: Dockerfile.py3 [hunk 7]
```diff
@@ -122,8 +122,8 @@ RUN useradd --create-home --shell /bin/bash ${USER}
 RUN usermod -append --groups sudo ${USER}
 RUN echo "%sudo ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
 
-
-RUN pip3 install --upgrade cython==0.28.6
+# install cython for python 2 (for python 3 it's inside the venv)
+pip2 install --upgrade Cython==0.28.6
```
Comment (AndreMiras) [inline Dockerfile.py3]: You're missing the `RUN` here:
```
Error response from daemon: Dockerfile parse error line 126: unknown instruction: PIP2
```
ahah don't give up :sweat_smile:

**Comment by AndreMiras** [comment]
> even though I believe this is the "more correct" way to call it in any case

Yes same feeling here, so we should make it into the tree in any case (after the build passes of course)

File: Dockerfile.py3 (hunk 7 above)
Comment (JonasT) [reply Dockerfile.py3]: > ahah don't give up :sweat_smile:

I won't I'm stubborn :joy:

**Comment by AndreMiras** [comment]
Good job! Feel free to merge once the build turns :heavy_check_mark:
